### PR TITLE
Fix ConcurrentModificationException in HostRegexTableLoadBalancer

### PR DESCRIPTION
**Pull request by adamjshook**

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java [hunk 1]
```diff
@@ -226,10 +227,10 @@ protected void parseConfiguration(ServerConfiguration conf) {
     if (null == t) {
       throw new RuntimeException("Table Operations cannot be null");
     }
-    tableIdToTableName = new HashMap<>();
-    poolNameToRegexPattern = new HashMap<>();
+    Map<String,String> tableIdToTableNameBuilder = new HashMap<>();
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java]: Could use an ImmutableMap.Builder

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java [hunk 2]
```diff
@@ -244,11 +245,15 @@ protected void parseConfiguration(ServerConfiguration conf) {
...
             String tableName = customProp.getKey().substring(HOST_BALANCER_PREFIX.length());
             String regex = customProp.getValue();
-            poolNameToRegexPattern.put(tableName, Pattern.compile(regex));
+            poolNameToRegexPatternBuilder.put(tableName, Pattern.compile(regex));
           }
         }
       }
     }
+
+    tableIdToTableName = ImmutableMap.copyOf(tableIdToTableNameBuilder);
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java]: Not completely sure, but I am wondering if tableIdToTableName and poolNameToRegexPattern should be made volatile.  Before this change the reference to the map never changed.  Now the ref is being changed to a new map.  Making it volatile would ensure other threads see the new map.

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java (hunk 1 above)
Comment (adamjshook) [reply server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java]: I tried that originally, but it contains nulls?  Or maybe it was duplicate keys.  I patched this a while ago but failed to contribute it back at the time.

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java (hunk 2 above)
Comment (adamjshook) [reply server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java]: Good catch; I will update it.

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java (hunk 1 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java]: I have run into those issues with Guava's immutable code before.  Your change seems like the best approach with those issues.

**Comment by adamjshook** [comment]
Thank you for the review @keith-turner!

**Comment by adamjshook** [comment]
@keith-turner Remind me what protocol here is in moving it upstream to `master`? It doesn't directly apply -- requires a minor modification to key by `TableId` instead of `String`.  Should I open another PR? Or cherry pick, update, and push straight to `master` with this PR number in the commit?

**Comment by keith-turner** [comment]
@adamjshook need to merge 1.9 into master and resolve conflicts.  I was actually working through that.  Would you like to do it?  Your commit and another commit needed to be merged and both had conflicts.

**Comment by adamjshook** [comment]
@keith-turner Ah, since you're already doing it, then please continue.  Thanks for the help.

**Comment by keith-turner** [comment]
I do not do PR for the merge commit.  Adding the PR issue number(s) to the merge commit would be nice.  I sometimes do this.

**Comment by keith-turner** [comment]
These are the changes before and after my merge.  Do these look good? 

https://github.com/apache/accumulo/compare/18320ad..1a373f5

**Comment by adamjshook** [comment]
Looks good to me!

**Pull request merged**